geo/src/algorithm/contains/polygon.rs

```diff
@@ -0,0 +1,91 @@
+use super::Contains;
+use crate::intersects::Intersects;
+use crate::kernels::*;
+use crate::*;
+
+// ┌─────────────────────────────┐
+// │ Implementations for Polygon │
+// └─────────────────────────────┘
+
+impl<T> Contains<Coordinate<T>> for Polygon<T>
+where
+    T: HasKernel,
+{
+    fn contains(&self, coord: &Coordinate<T>) -> bool {
+        match utils::coord_pos_relative_to_ring(*coord, &self.exterior()) {
+            utils::CoordPos::OnBoundary | utils::CoordPos::Outside => false,
+            _ => self.interiors().iter().all(|ls| {
+                utils::coord_pos_relative_to_ring(*coord, ls) == utils::CoordPos::Outside
+            }),
+        }
+    }
+}
+
+impl<T> Contains<Point<T>> for Polygon<T>
+where
+    T: HasKernel,
+{
+    fn contains(&self, p: &Point<T>) -> bool {
+        self.contains(&p.0)
+    }
+}
+
+// TODO: ensure DE-9IM compliance: esp., when
+// line.start and line.end is on the boundaries
+impl<T> Contains<Line<T>> for Polygon<T>
+where
+    T: HasKernel,
+{
+    fn contains(&self, line: &Line<T>) -> bool {
+        // both endpoints are contained in the polygon and the line
+        // does NOT intersect the exterior or any of the interior boundaries
+        self.contains(&line.start)
+            && self.contains(&line.end)
+            && !self.exterior().intersects(line)
+            && !self.interiors().iter().any(|inner| inner.intersects(line))
+    }
+}
+
+// TODO: also check interiors
+impl<T> Contains<Polygon<T>> for Polygon<T>
+where
+    T: HasKernel,
+{
+    fn contains(&self, poly: &Polygon<T>) -> bool {
+        // decompose poly's exterior ring into Lines, and check each for containment
+        poly.exterior().lines().all(|line| self.contains(&line))
+    }
+}
+
+// TODO: ensure DE-9IM compliance
+impl<T> Contains<LineString<T>> for Polygon<T>
+where
+    T: HasKernel,
+{
+    fn contains(&self, linestring: &LineString<T>) -> bool {
+        // All LineString points must be inside the Polygon
+        if linestring.points_iter().all(|point| self.contains(&point)) {
+            // The Polygon interior is allowed to intersect with the LineString
+            // but the Polygon's rings are not
+            !self
+                .interiors()
+                .iter()
+                .any(|ring| ring.intersects(linestring))
+        } else {
+            false
+        }
+    }
+}
+
+// ┌──────────────────────────────────┐
+// │ Implementations for MultiPolygon │
+// └──────────────────────────────────┘
+impl<G, T> Contains<G> for MultiPolygon<T>
+where
+    T: CoordinateType,
+    Polygon<T>: Contains<G>,
```

Comment:
It looks like we're newly accepting *any* type which implements the Contains trait, whereas before MultiPolygon only supported Coordinate and Point, is that right?

Does this existing implementation work for types with more dimensionality than a point?

e.g. If the two squares are part of a single multipolygon, even though the black line is contained by neither square individually, is it contained by the multipolygon?

![image](https://user-images.githubusercontent.com/217057/93694257-6432c300-fabe-11ea-9d4d-2d4474ab9180.png)

Reply:
So, [MultiPolygon](https://locationtech.github.io/jts/javadoc/org/locationtech/jts/geom/MultiPolygon.html) should not contain overlapping constituent polys: 

>As per the OGC SFS specification, the Polygons in a MultiPolygon may not overlap, and may only touch at single points. This allows the topological point-set semantics to be well-defined.

With the above defn., I think we can lift any Polygon: Contains<T> to work on MultiPolygon